File: client/src/utils/number-utils.ts
import {isDefined} from "./object-utils";

interface Bound {
    value?: number,
    mode: "inclusive" | "exclusive"
}

interface IsBetweenOptions {
    lowerBound: Bound,
    upperBound: Bound
}

export const isBetween = (number: number, bounds: IsBetweenOptions): boolean => {
    const {lowerBound, upperBound} = bounds;

    if (!isDefined(lowerBound.value) && !isDefined(upperBound.value)) {
        return true;
    }

    const satisfiesLowerBound = isDefined(lowerBound.value)
        ? lowerBound.mode === "inclusive" ? number >= lowerBound.value! : number > lowerBound.value!
        : true;
    const satisfiesUpperBound = isDefined(upperBound.value)
        ? upperBound.mode === "inclusive" ? number <= upperBound.value! : number < upperBound.value!
        : true;

    return satisfiesLowerBound && satisfiesUpperBound;
}

File: chat/src/main/kotlin/chatox/chat/util/NumberUtils.kt
package chatox.chat.util

enum class BoundMode {
    INCLUSIVE,
    EXCLUSIVE
}

data class Bound(val value: Int?, val mode: BoundMode)

fun isBetween(number: Int, lowerBound: Bound, upperBound: Bound): Boolean {
    if (lowerBound.value == null && upperBound.value == null) {
        return true
    }

    val satisfiesLowerBound = if (lowerBound.value != null) {
        if (lowerBound.mode == BoundMode.INCLUSIVE) number >= lowerBound.value else number > lowerBound.value
    } else {
        true
    }
    val satisfiedUpperBound = if (upperBound.value != null) {
        if (upperBound.mode == BoundMode.INCLUSIVE) number <= upperBound.value else number < upperBound.value
    } else {
        true
    }

    return satisfiesLowerBound && satisfiedUpperBound
}